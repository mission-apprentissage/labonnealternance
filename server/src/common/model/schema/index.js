import { apiCallSchema } from "./ApiCall.js"
import { applicationSchema } from "./applications.js"
import { appointmentSchema } from "./appointment.js"
import { credentialSchema } from "./credentials.js"
import { diplomesMetiersSchema } from "./diplomesmetiers.js"
import { domainesMetiersSchema } from "./domainesmetiers.js"
import { etablissementSchema } from "./etablissement.js"
import { formulaireSchema } from "./formulaire.js"
import { bonneBoiteSchema } from "./lbb/bonneBoite.js"
import { emailBlacklist } from "./lbb/emailBlacklist.js"
import { geoLocationSchema } from "./lbb/geoLocations.js"
import { opco } from "./lbb/opco.js"
import { mnaFormationSchema } from "./mnaFormation/mnaFormation.js"
import { offreSchema } from "./offre.js"
import { optoutSchema } from "./optout.js"
import { parcoursupEtablissementStatSchema } from "./parcoursupEtablissementStat.js"
import { userRecruteurSchema } from "./recruteur/users.js"
import { referentielOpcoSchema } from "./referentielOpco.js"
import { userSchema } from "./user.js"
import { widgetParameterSchema } from "./widgetParameter.js"

export default {
  domainesMetiersSchema,
  diplomesMetiersSchema,
  applicationSchema,
  apiCallSchema,
  mnaFormationSchema,
  bonneBoiteSchema,
  geoLocationSchema,
  emailBlacklist,
  opco,
  userSchema,
  appointmentSchema,
  widgetParameterSchema,
  etablissementSchema,
  parcoursupEtablissementStatSchema,
  formulaireSchema,
  offreSchema,
  referentielOpcoSchema,
  optoutSchema,
  userRecruteurSchema,
  credentialSchema,
}
