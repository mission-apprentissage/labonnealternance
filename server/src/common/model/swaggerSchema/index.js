import requireAll from "require-all"
import __dirname from "../../dirname.js"

let models = {}

function filterFile(filename) {
  if (filename.endsWith(".js") && filename !== "index.js") {
    return filename
  }

  return false
}

const modelsList = requireAll({
  dirname: __dirname(import.meta.url),
  filter: (filename) => filterFile(filename),
})

Object.keys(modelsList).forEach((filename) => {
  const model = modelsList[filename]
  models = { ...models, ...model }
})

export default models
