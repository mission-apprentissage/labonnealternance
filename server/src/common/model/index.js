import mongoosePaginate from "mongoose-paginate"
import { getElasticInstance, mongoosastic } from "../esClient/index.js"
import schema from "../model/schema/index.js"
import { mongooseInstance } from "../mongodb.js"

const createModel = (modelName, descriptor, options = {}) => {
  const schema = new mongooseInstance.Schema(descriptor)

  if (options.esIndexName) {
    schema.plugin(mongoosastic, { esClient: getElasticInstance(), index: options.esIndexName })
  }

  if (options.esIndexName || options.paginate) {
    schema.plugin(mongoosePaginate)
  }

  if (options.createMongoDBIndexes) {
    options.createMongoDBIndexes(schema)
  }

  return mongooseInstance.model(modelName, schema)
}

const DomainesMetiers = createModel("domainesmetiers", schema.domainesMetiersSchema, {
  esIndexName: "domainesmetiers",
})
const FormationCatalogue = createModel("formationcatalogues", schema.mnaFormationSchema, {
  esIndexName: "formationcatalogues",
})
const DiplomesMetiers = createModel("diplomesmetiers", schema.diplomesMetiersSchema, {
  esIndexName: "diplomesmetiers",
})
const ApiCalls = createModel("apicalls", schema.apiCallSchema)
const Application = createModel("applications", schema.applicationSchema, {
  paginate: true,
})
const SourceFormations = createModel("sourceformations", schema.sourceFormationsSchema)
const GeoLocation = createModel("geolocation", schema.geoLocationSchema)
const EmailBlacklist = createModel("emailblacklist", schema.emailBlacklist)
const Opco = createModel("opco", schema.opco)
const BonnesBoites = createModel("bonnesboites", schema.bonneBoiteSchema, {
  esIndexName: "bonnesboites",
})
const User = createModel("user", schema.userSchema, { paginate: true })
const Appointment = createModel("appointment", schema.appointmentSchema, { paginate: true })
const WidgetParameter = createModel("widgetParameter", schema.widgetParameterSchema, { paginate: true })
const Etablissement = createModel("etablissement", schema.etablissementSchema, { paginate: true })
const ParcoursupEtablissementStat = createModel("parcoursupEtablissementStat", schema.parcoursupEtablissementStatSchema, { paginate: true })
const Formulaire = createModel("formulaire", schema.formulaireSchema, { esIndexName: "formulaires", paginate: true })
const Credential = createModel("credentials", schema.credentialSchema)
const Offre = createModel("offre", schema.offresSchema)
const Optout = createModel("optout", schema.optoutSchema, { paginate: true })
const ReferentielOpco = createModel("referentielOpco", schema.referentielOpcoSchema)
const UserRecruteur = createModel("userRecruteur", schema.userRecruteurSchema, { paginate: true })
const AppointmentDetailed = createModel("appointmentDetailed", schema.appointmentSchema)

export {
  DomainesMetiers,
  FormationCatalogue,
  DiplomesMetiers,
  ApiCalls,
  Application,
  SourceFormations,
  GeoLocation,
  EmailBlacklist,
  Opco,
  BonnesBoites,
  User,
  Appointment,
  WidgetParameter,
  Etablissement,
  ParcoursupEtablissementStat,
  Formulaire,
  Credential,
  Offre,
  Optout,
  ReferentielOpco,
  UserRecruteur,
  AppointmentDetailed,
}
