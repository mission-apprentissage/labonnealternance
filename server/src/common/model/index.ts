import { captureException } from "@sentry/node"

import { db, mongooseInstance } from "@/common/mongodb"

import { logger } from "../logger"

import AnonymizedUser from "./schema/anonymizedUsers/anonymizedUsers.schema"
import ApiCalls from "./schema/apiCall/apiCall.schema"
import AnonymizedApplication from "./schema/application/anonymizedApplications.schema"
import Application from "./schema/application/applications.schema"
import AppointmentDetailed from "./schema/appointmentDetailed/appointmentDetailed.schema"
import Appointment from "./schema/appointments/appointment.schema"
import Credential from "./schema/credentials/credential.schema"
import CustomEmailETFA from "./schema/customEmailETFA/customEmailETFA.schema"
import DiplomesMetiers from "./schema/diplomesmetiers/diplomesmetiers.schema"
import DomainesMetiers from "./schema/domainesmetiers/domainesmetiers.schema"
import EligibleTrainingsForAppointment from "./schema/eligibleTrainingsForAppointment/eligibleTrainingsForAppointment.schema"
import eligibleTrainingsForAppointmentHistory from "./schema/eligibleTrainingsForAppointmentsHistory/eligibleTrainingsForAppointmentHistory.schema"
import EmailBlacklist from "./schema/emailBlacklist/emailBlacklist.schema"
import Etablissement from "./schema/etablissements/etablissement.schema"
import FicheMetierRomeV4 from "./schema/ficheRomeV4/ficheRomeV4"
import FormationCatalogue from "./schema/formationCatalogue/formationCatalogue.schema"
import GeoLocation from "./schema/geolocation/geolocation.schema"
import InternalJobs from "./schema/internalJobs/internalJobs.schema"
import Job from "./schema/jobs/jobs.schema"
import LbaCompany from "./schema/lbaCompany/lbaCompany.schema"
import LbaCompanyLegacy from "./schema/lbaCompanylegacy/lbaCompanyLegacy.schema"
import { Cfa } from "./schema/multiCompte/cfa.schema"
import { Entreprise } from "./schema/multiCompte/entreprise.schema"
import { RoleManagement } from "./schema/multiCompte/roleManagement.schema"
import { UserWithAccount } from "./schema/multiCompte/userWithAccount.schema"
import Opco from "./schema/opco/opco.schema"
import Optout from "./schema/optout/optout.schema"
import Recruiter from "./schema/recruiter/recruiter.schema"
import RecruteurLbaUpdateEvent from "./schema/recruteurLbaUpdateEvent/recruteurLbaUpdateEvent.schema"
import ReferentielOnisep from "./schema/referentielOnisep/referentielOnisep.schema"
import ReferentielOpco from "./schema/referentielOpco/referentielOpco.schema"
import ReferentielRome from "./schema/referentielRome/referentielRome"
import Session from "./schema/session/session.schema"
import SiretDiffusibleStatus from "./schema/siretDiffusibleStatusSchema/siretDiffusibleStatusSchema.schema"
import UnsubscribedLbaCompany from "./schema/unsubscribedLbaCompany/unsubscribedLbaCompany.schema"
import UnsubscribeOF from "./schema/unsubscribedOF/unsubscribeOF.schema"
import User from "./schema/user/user.schema"

const createSpecialIndexes = async () => {
  await db.collection("bonnesboites").createIndex({ geopoint: "2dsphere" })
  await db.collection("formationcatalogues").createIndex({ lieu_formation_geopoint: "2dsphere" })
  await db.collection("recruiters").createIndex({ geopoint: "2dsphere" })
}

export async function createMongoDBIndexes() {
  const results = await Promise.allSettled(
    mongooseInstance.modelNames().map(async (name) => {
      const model = mongooseInstance.model(name)
      return model.createIndexes({ background: true }).catch(async (e) => {
        if (e.codeName === "IndexOptionsConflict" || e.codeName === "IndexKeySpecsConflict") {
          const err = new Error(`Conflict in indexes for ${name}`, { cause: e })
          logger.error(err)
          captureException(err)
          await mongooseInstance.connection.collection(model.collection.name).dropIndexes()
          await model.createIndexes({ background: true })
        } else {
          throw e
        }
      })
    })
  )

  await createSpecialIndexes()

  const errors = results.reduce((acc, r) => {
    if (r.status === "rejected") {
      acc.push(r.reason)

      logger.error(r.reason)
      captureException(r.reason)
    }

    return acc
  }, [] as Error[])

  if (errors.length > 0) {
    throw new AggregateError(errors, `createMongoDBIndexes failed with ${errors.length} errors`)
  }
}

export {
  AnonymizedApplication,
  AnonymizedUser,
  ApiCalls,
  Application,
  Appointment,
  AppointmentDetailed,
  Cfa,
  Credential,
  CustomEmailETFA,
  DiplomesMetiers,
  DomainesMetiers,
  EligibleTrainingsForAppointment,
  EmailBlacklist,
  Entreprise,
  Etablissement,
  FicheMetierRomeV4,
  FormationCatalogue,
  GeoLocation,
  InternalJobs,
  Job,
  LbaCompany,
  LbaCompanyLegacy,
  RecruteurLbaUpdateEvent,
  Opco,
  Optout,
  Recruiter,
  ReferentielOnisep,
  ReferentielOpco,
  ReferentielRome,
  RoleManagement,
  Session,
  SiretDiffusibleStatus,
  UnsubscribeOF,
  UnsubscribedLbaCompany,
  User,
  UserWithAccount,
  eligibleTrainingsForAppointmentHistory,
}
