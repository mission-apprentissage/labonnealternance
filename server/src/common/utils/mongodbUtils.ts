import { captureException } from "@sentry/node"
import { isEqual } from "lodash-es"
import { Collection, CollectionInfo, MongoClient, MongoServerError } from "mongodb" // to uncomment when migrated to V7
import { IModelDescriptor } from "shared/models/common"
import { CollectionName, IDocument, modelDescriptors } from "shared/models/models"
import { zodToMongoSchema } from "zod-mongodb-schema"

import config from "../../config"
import { logger } from "../logger"

import { sleep } from "./asyncUtils"

let mongodbClient: MongoClient | null = null
let mongodbClientState: string | null = null

export const ensureInitialization = () => {
  if (!mongodbClient) {
    throw new Error("Database connection does not exist. Please call connectToMongodb before.")
  }
  return mongodbClient
}

/**
 * @param  {string} uri
 * @returns client
 */
export const connectToMongodb = async (uri: string) => {
  const client = new MongoClient(uri, {
    heartbeatFrequencyMS: 10_000,
    retryWrites: true,
    retryReads: true,
    minPoolSize: config.env === "local" ? 0 : 5,
    maxPoolSize: 1_000,
  })

  client.on("connectionPoolReady", () => {
    logger.info("MongoDB reconnected")
    mongodbClient = client
  })

  client.on("connectionPoolClosed", () => {
    logger.warn("MongoDB closed")
    mongodbClient = null
  })

  await client.connect()
  // @ts-expect-error
  mongodbClientState = client.topology.s.state
  mongodbClient = client
  logger.info("Connected to MongoDB")

  return client
}

export const getMongodbClient = () => mongodbClient
export const getMongodbClientState = () => mongodbClientState

export const closeMongodbConnection = async () => {
  logger.warn("Closing MongoDB")
  // Let 100ms for possible callback cleanup to register tasks in mongodb queue
  await sleep(200)
  return mongodbClient?.close()
}

export const getDatabase = () => {
  return ensureInitialization().db()
}

export const getDbCollection = <K extends CollectionName>(name: K): Collection<IDocument<K>> => {
  return ensureInitialization().db().collection(name)
}

export const getCollectionList = () => {
  return ensureInitialization().db().listCollections().toArray()
}

export const getDbCollectionIndexes = async (name: CollectionName) => {
  return await ensureInitialization().db().collection(name).indexes()
}

/**
 * Création d'une collection si elle n'existe pas
 * @param {string} collectionName
 */
const createCollectionIfDoesNotExist = async (collectionName: string) => {
  const db = getDatabase()
  const collectionsInDb = await db.listCollections().toArray()
  const collectionExistsInDb = collectionsInDb.map(({ name }) => name).includes(collectionName)

  if (!collectionExistsInDb) {
    try {
      await db.createCollection(collectionName)
    } catch (err) {
      if ((err as MongoServerError).codeName !== "NamespaceExists") {
        throw err
      }
    }
  }
}

/**
 * Vérification de l'existence d'une collection à partir de la liste des collections
 * @param {*} collectionsInDb
 * @param {*} collectionName
 * @returns
 */
export const collectionExistInDb = (collectionsInDb: CollectionInfo[], collectionName: string) => collectionsInDb.map(({ name }: { name: string }) => name).includes(collectionName)

/**
 * Config de la validation
 * @param {*} modelDescriptors
 */
export const configureDbSchemaValidation = async (modelDescriptors: IModelDescriptor[]) => {
  const db = getDatabase()
  ensureInitialization()
  await Promise.all(
    modelDescriptors.map(async ({ collectionName, zod }) => {
      await createCollectionIfDoesNotExist(collectionName)

      const convertedSchema = zodToMongoSchema(zod)

      try {
        await db.command({
          collMod: collectionName,
          validationLevel: "strict",
          validationAction: "error",
          validator: {
            $jsonSchema: {
              title: `${collectionName} validation schema`,
              ...convertedSchema,
            },
          },
        })
      } catch (error) {
        captureException(error)
        logger.error(error)
      }
    })
  )
}

/**
 * Clear de toutes les collections
 * @returns
 */
export const clearAllCollections = async () => {
  const collections = await getDatabase().collections()
  return Promise.all(collections.map((c) => c.deleteMany({})))
}

/**
 * Clear d'une collection
 * @param {string} name
 * @returns
 */
export async function clearCollection(name: string) {
  ensureInitialization()
  await getDatabase().collection(name).deleteMany({})
}

export const createIndexes = async () => {
  for (const descriptor of modelDescriptors) {
    if (!descriptor.indexes) {
      return
    }
    const indexes = await getDbCollection(descriptor.collectionName)
      .listIndexes()
      .toArray()
      .catch((err) => {
        // NamespaceNotFound
        if (err.code === 26) {
          return []
        }
        throw err
      })
    const indexesToRemove = new Set(indexes.filter((i) => i.name !== "_id_"))

    logger.info(`Create indexes for collection ${descriptor.collectionName}`)
    await Promise.all(
      descriptor.indexes.map(async ([index, options]): Promise<void> => {
        try {
          const existingIndex =
            // Use Object.entries because order matters
            indexes.find((i) => isEqual(Object.entries(i.key), Object.entries(index)) || options.name === i.name) ?? null

          if (existingIndex) {
            indexesToRemove.delete(existingIndex)
          }

          await getDbCollection(descriptor.collectionName)
            .createIndex(index, options)
            .catch(async (err) => {
              // IndexOptionsConflict & IndexKeySpecsConflict
              if (err.code === 85 || err.code === 86) {
                await getDbCollection(descriptor.collectionName).dropIndex(existingIndex.name)
                await getDbCollection(descriptor.collectionName).createIndex(index, options)
              } else {
                throw err
              }
            })
        } catch (err) {
          captureException(err)
          logger.error(`Error creating indexes for ${descriptor.collectionName}: ${err}`)
        }
      })
    )

    if (indexesToRemove.size > 0) {
      logger.warn(`Dropping extra indexes for collection ${descriptor.collectionName}`, indexesToRemove)
      await Promise.all(Array.from(indexesToRemove).map((index) => getDbCollection(descriptor.collectionName).dropIndex(index.name)))
    }
  }
}

export const dropIndexes = async () => {
  const collections = (await getCollectionList()).map((collection) => collection.name)
  for (const descriptor of modelDescriptors) {
    logger.info(`Drop indexes for collection ${descriptor.collectionName}`)
    if (collections.includes(descriptor.collectionName)) {
      await getDbCollection(descriptor.collectionName).dropIndexes()
    }
  }
}
