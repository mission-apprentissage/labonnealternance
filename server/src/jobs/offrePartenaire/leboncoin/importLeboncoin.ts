import { JOBPARTNERS_LABEL } from "shared/models/jobsPartners.model"
import rawLeboncoinModel from "shared/models/rawLeboncoin.model"

import config from "@/config"
import { importFromStreamInCsv } from "@/jobs/offrePartenaire/importFromStreamInCsv"
import { importFromUrlInCsv } from "@/jobs/offrePartenaire/importFromUrlInCsv"

import { rawToComputedJobsPartners } from "../rawToComputedJobsPartners"

import { leboncoinJobToJobsPartners, ZLeboncoinJob } from "./leboncoinMapper"

const rawCollectionName = rawLeboncoinModel.collectionName
const documentJobRoot = "job"

export const importLeboncoin = async (sourceStream?: NodeJS.ReadableStream) => {
  if (sourceStream) {
    await importFromStreamInCsv({
      destinationCollection: rawCollectionName,
      stream: sourceStream,
      partnerLabel: JOBPARTNERS_LABEL.LEBONCOIN,
      parseOptions: { delimiter: "," },
    })
  } else {
    await importFromUrlInCsv({
      destinationCollection: rawCollectionName,
      url: config.leboncoinUrl,
      partnerLabel: JOBPARTNERS_LABEL.LEBONCOIN,
      parseOptions: { delimiter: "," },
    })
  }
}

export const importLeboncoinToComputed = async () => {
  await rawToComputedJobsPartners({
    collectionSource: rawCollectionName,
    partnerLabel: JOBPARTNERS_LABEL.LEBONCOIN,
    zodInput: ZLeboncoinJob,
    mapper: leboncoinJobToJobsPartners,
    documentJobRoot,
  })
}
