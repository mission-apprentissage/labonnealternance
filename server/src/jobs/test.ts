import { ObjectId } from "bson"
import { IApplicant, IApplication, ZApplicant } from "shared"

import { getDbCollection } from "@/common/utils/mongodbUtils"

import { asyncForEach } from "../common/utils/asyncUtils"

type ApplicationAggregate = Pick<IApplication, "applicant_email" | "applicant_first_name" | "applicant_last_name" | "applicant_phone" | "_id">

export const migrationTest = async () => {
  // console.log(extensions.telephone.safeParse("0690760552"))
  // return
  const applications = (await getDbCollection("applications")
    .find({}, { projection: { applicant_first_name: 1, applicant_last_name: 1, applicant_email: 1, applicant_phone: 1, _id: 1 } })
    .toArray()) as ApplicationAggregate[]

  const stat = { error: 0, success: 0, total: applications.length }

  await asyncForEach(applications, async (application) => {
    const { applicant_first_name, applicant_email, applicant_last_name, applicant_phone, _id } = application
    const now = new Date()
    const applicant: IApplicant = {
      _id: new ObjectId(),
      firstname: applicant_first_name,
      lastname: applicant_last_name,
      phone: applicant_phone,
      email: applicant_email,
      last_connection: null,
      createdAt: now,
      updatedAt: now,
    }
    const validation = ZApplicant.safeParse(applicant)
    if (!validation.success) {
      stat.error++
      console.log({ applicant, err: validation.error.errors })
    } else {
      stat.success++
    }
  })
  console.log(stat)
}
