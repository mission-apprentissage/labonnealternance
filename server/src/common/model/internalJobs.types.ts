import { ObjectId } from "mongodb"

import { CronName } from "@/jobs/jobs"

/*
  name: Le nom de la tâche
  type: Type du job simple ou cron
  status: Statut courant du job
  sync: Si le job est synchrone
  payload: La donnée liéé à la tâche
  output: Les valeurs de retours du job
  cron_string: standard cron string exemple: '*\/2 * * * *'
  scheduled_for: Date de lancement programmée
  started_at: Date de lancement
  ended_at: Date de fin d'execution
  updated_at: Date de mise à jour en base de données
  created_at: Date d'ajout en base de données
*/

interface IInternalJobsSimple {
  _id: ObjectId
  name: string
  type: "simple"
  status: "pending" | "will_start" | "running" | "finished" | "blocked" | "errored"
  sync: boolean
  payload: Record<string, any>
  output?: Record<string, unknown> | undefined
  scheduled_for: Date
  started_at?: Date
  ended_at?: Date
  updated_at: Date
  created_at: Date
}

interface IInternalJobsCron {
  _id: ObjectId
  name: CronName
  type: "cron"
  status: "pending" | "will_start" | "running" | "finished" | "blocked" | "errored"
  sync: boolean
  cron_string: string
  scheduled_for: Date
  updated_at: Date
  created_at: Date
}

interface IInternalJobsCronTask {
  _id: ObjectId
  name: CronName
  type: "cron_task"
  status: "pending" | "will_start" | "running" | "finished" | "blocked" | "errored"
  sync: boolean
  scheduled_for: Date
  started_at?: Date
  ended_at?: Date
  updated_at: Date
  created_at: Date
}

type IInternalJobs = IInternalJobsSimple | IInternalJobsCron | IInternalJobsCronTask

export type { CronName, IInternalJobs, IInternalJobsCron, IInternalJobsCronTask, IInternalJobsSimple }
