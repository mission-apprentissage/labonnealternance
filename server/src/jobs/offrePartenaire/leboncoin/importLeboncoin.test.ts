import fs from "node:fs"

import { JOBPARTNERS_LABEL } from "shared/models/jobsPartners.model"
import { beforeEach, describe, expect, it, vi } from "vitest"

import { getDbCollection } from "@/common/utils/mongodbUtils"
import { useMongo } from "@tests/utils/mongo.test.utils"

import { importLeboncoin, importLeboncoinToComputed } from "./importLeboncoin"

const now = new Date("2024-07-21T04:49:06.000+02:00")

describe("importLeboncoin", () => {
  useMongo()

  beforeEach(() => {
    vi.useFakeTimers()
    vi.setSystemTime(now)

    return async () => {
      vi.useRealTimers()
      await getDbCollection("computed_jobs_partners").deleteMany({})
      await getDbCollection("raw_leboncoin").deleteMany({})
    }
  })

  it("should test the import of Le bon coin emploi data into computed_job_partners", async () => {
    const fileStream = fs.createReadStream("server/src/jobs/offrePartenaire/leboncoin/importLeboncoin.test.input.csv")
    await importLeboncoin(fileStream)
    expect.soft(await getDbCollection("raw_leboncoin").countDocuments({})).toBe(2)

    await importLeboncoinToComputed()
    const jobs = (
      await getDbCollection("computed_jobs_partners")
        .find({ partner_label: JOBPARTNERS_LABEL.LEBONCOIN }, { projection: { _id: 0, created_at: 0 } })
        .toArray()
    ).sort((a, b) => ((a.partner_job_id ?? "") < (b.partner_job_id ?? "") ? -1 : 1))
    expect.soft(jobs.length).toBe(2)
    expect.soft(jobs).toMatchSnapshot()
  })
})
