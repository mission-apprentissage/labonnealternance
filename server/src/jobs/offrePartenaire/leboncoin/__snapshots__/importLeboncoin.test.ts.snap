// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`importLeboncoin > should test the import of Le bon coin emploi data into computed_job_partners 1`] = `
[
  {
    "apply_phone": null,
    "apply_url": "https://www.leboncoin.fr/offres_d_emploi/2999583187.htm",
    "business_error": null,
    "contract_duration": null,
    "contract_is_disabled_elligible": false,
    "contract_remote": null,
    "contract_start": null,
    "contract_type": [],
    "errors": [],
    "jobs_in_success": [],
    "lba_url": null,
    "offer_access_conditions": [],
    "offer_creation": null,
    "offer_description": "R√©sum√©
ALTERNANCE - Boucher F/H, 37540 Saint Cyr Sur Loire, Alternance, Temps plein.

Description du poste
‚ú®Nous sommes‚Ä¶¬†
Auchan Retail France, une entreprise de distribution de plus de 50 000 collaborateurs au service de millions de clients chaque ann√©e.
Ce que nous voulons, c‚Äôest aider nos clients √† bien manger et vivre mieux, tout en pr√©servant la plan√®te.
Que l‚Äôon soit en entrep√¥t, en magasin ou sur des fonctions support, notre point commun √† tous, c‚Äôest de faire en sorte que nos clients aient plaisir √† faire leurs courses chez nous.
Parce que nos collaborateurs sont aussi nos premiers clients, nous veillons chaque jour √† leur bien-√™tre et leur √©volution professionnelle.Ensemble, nous am√©liorons le quotidien !
üìåVous avez envie de‚Ä¶
Connaitre, transformer les produits de votre march√©. Vous mettez en ≈ìuvre les d√©coupes et recettes / process de transformation de vos produits. Vous √™tes polyvalent √† tous les postes de fabrication / d√©coupe avec un haut niveau de professionnalisme. Vous √©valuez la qualit√© et la fra√Æcheur des produits et d'√©carter les produits non-conformes.

Contribuer √† la bonne organisation de la production dans le respect des r√®gles d'hygi√®ne. ¬†Vous organisez votre espace de travail (propret√©, rangement, entretien du mat√©riel) et veillez au respect des r√®gles de s√©curit√© alimentaire. Enfin, vous √©talez la fabrication afin de garantir la fra√Æcheur des produits et de r√©pondre √† l'objectif de z√©ro rupture.¬†

Mettre en valeurs vos produits et fid√©liser vos clients. Vous approvisionnez et mettez en avant vos produits, dans le respect des r√®gles d'√©talagisme et de merchandising. Vous veillez √† la mise en valeur de vos produits, dans le but d'attirer et fid√©liser vos clients. Vous conseillez et r√©pondez aux attentes de ces derniers.¬†
üéØVous √™tes‚Ä¶En formation de niveau BP/CAP/BAC PRO en boucherie ou un CQP. Vous aimez les produits issu de ce domaine.Le poste de Boucher en alternance F/H est fait pour vous.Rencontrons-nous !




Profil recherch√©

üì¢Nous vous offrons les avantages suivants :


‚Ä¢ Une remise sur achat chez Auchan
‚Ä¢ Des primes de participation et d‚Äôint√©ressement selon la performance annuelle de l‚Äôentreprise
‚Ä¢ Une prime annuelle √©quivalente √† un 13e mois


Chez Auchan, nous sommes convaincus que la diversit√© fait la richesse d‚Äôune entreprise. Nous √©tudions √† comp√©tences √©gales chaque candidature et toutes nos offres peuvent faire l‚Äôobjet d‚Äôam√©nagements sp√©cifiques en cas de handicap - #tous √©gaux, tous diff√©rents !


üå≥Vous voulez en savoir plus sur nos engagements ?
Nous nous mobilisons au quotidien autour de 4 piliers de responsabilit√© soci√©tale et environnementale majeurs : Offre responsable, Environnement, Solidarit√©, Humain.Rendez-vous sur : [URL masqu√©e pour votre s√©curit√©].",
    "offer_desired_skills": [],
    "offer_expiration": 2024-09-21T02:49:06.000Z,
    "offer_multicast": true,
    "offer_opening_count": 1,
    "offer_origin": null,
    "offer_rome_codes": null,
    "offer_status": "Active",
    "offer_status_history": [],
    "offer_target_diploma": null,
    "offer_title": "ALTERNANCE - Boucher F/H",
    "offer_to_be_acquired_knowledge": [],
    "offer_to_be_acquired_skills": [],
    "partner_job_id": "2999583187",
    "partner_label": "Le bon coin emploi",
    "updated_at": 2024-07-21T02:49:06.000Z,
    "validated": false,
    "workplace_address_city": "Saint-Cyr-sur-Loire",
    "workplace_address_label": "Saint-Cyr-sur-Loire 37540",
    "workplace_address_street_label": null,
    "workplace_address_zipcode": "37540",
    "workplace_brand": null,
    "workplace_description": null,
    "workplace_geopoint": null,
    "workplace_idcc": null,
    "workplace_legal_name": null,
    "workplace_naf_code": null,
    "workplace_naf_label": null,
    "workplace_name": "Auchan",
    "workplace_opco": null,
    "workplace_siret": null,
    "workplace_size": null,
    "workplace_website": null,
  },
  {
    "apply_phone": null,
    "apply_url": "https://www.leboncoin.fr/offres_d_emploi/3071721792.htm",
    "business_error": null,
    "contract_duration": null,
    "contract_is_disabled_elligible": false,
    "contract_remote": null,
    "contract_start": null,
    "contract_type": [],
    "errors": [],
    "jobs_in_success": [],
    "lba_url": null,
    "offer_access_conditions": [],
    "offer_creation": null,
    "offer_description": "R√©sum√©
Conseiller Commercial Apr√®s-Vente en Alternance (H/F), 8000 Charleville Mezieres, Contrat d'alternance (pro, apprentissage), Temps plein.
Fonction : Assistant-te Service APV
Entreprise : BMW CHARLEVILLE

Description du poste
Distributeur automobile multi-marques depuis plus de 30 ans, le Groupe GCA est le¬†1er distributeur¬†fran√ßais de v√©hicules hybrides.¬†Le groupe est repr√©sent√© par 15 marques automobiles, 2 marques de moto, 2 000 collaborateurs, 3 centres v√©hicules d‚Äôoccasion, 10 carrosseries, 4 magasins de pi√®ces d√©tach√©es et un centre d'appels.¬†Avec le d√©veloppement de nouvelles technologies et les diff√©rents enjeux environnementaux, le monde de l‚Äôautomobile est plus que jamais en changement.

Chez GCA, nous misons sur vos comp√©tences et vous accompagnons dans l‚Äô√©volution de votre m√©tier en vous apportant les outils n√©cessaires √† votre r√©ussite.

Par ailleurs, nous vous aidons dans la mise en place de vos souhaits de mobilit√©, qu‚Äôelle soit g√©ographique ou professionnelle. Au sein de notre r√©seau de 120 concessions r√©parties dans 7 r√©gions de France et en Belgique, vous trouverez forc√©ment une concession pour r√©aliser vos diff√©rents choix de carri√®re.

√Ä propos du poste

Lors de votre pr√©sence¬†sur site, vous serez en charge des encaissements et de la facturation ainsi que du suivi des dossiers sp√©cifiques li√©s aux activit√©s apr√®s-vente de la concession. De plus, vous assurerez l'accueil de la client√®le et des fournisseurs, la tenue du standard t√©l√©phonique, les prises de rendez-vous et le suivi du planning de l'atelier.¬†

N'h√©sitez plus et venez nous rejoindre au sein de notre concession BMW √† Charleville !

Profil recherch√©
Vous int√©grez une formation de type Assistant de Gestion et/ou commerce (Bac+2/3). Vous √™tes organis√©(e) dans votre travail. Une exp√©rience dans le secteur de l'automobile serait un plus.


- Alternance
- Mutuelle entreprise offrant divers avantages sociaux (abonnements sportifs, vacances, scolarit√© etc.)
- Int√©ressement aux b√©n√©fices de l'entreprise
- R√©mun√©ration n√©gociable selon profil et exp√©rience

Int√©grez un Groupe en pleine croissance, ou l'esprit d'√©quipe et la satisfaction clients font partie de notre ADN. Nous vous accompagnons dans votre √©volution et votre √©panouissement professionnel √† travers des formations d√©di√©es √† nos marques, des challenges, des voyages et activit√©s avec les collaborateurs GCA !¬†Retrouvez toutes nos actualit√©s sur Instagram : @gca_recrute",
    "offer_desired_skills": [],
    "offer_expiration": 2024-09-21T02:49:06.000Z,
    "offer_multicast": true,
    "offer_opening_count": 1,
    "offer_origin": null,
    "offer_rome_codes": null,
    "offer_status": "Active",
    "offer_status_history": [],
    "offer_target_diploma": null,
    "offer_title": "Conseiller Commercial Apr√®s-Vente en Alternance (H/F)",
    "offer_to_be_acquired_knowledge": [],
    "offer_to_be_acquired_skills": [],
    "partner_job_id": "3071721792",
    "partner_label": "Le bon coin emploi",
    "updated_at": 2024-07-21T02:49:06.000Z,
    "validated": false,
    "workplace_address_city": "Charleville-M√©zi√®res",
    "workplace_address_label": "Charleville-M√©zi√®res 08000",
    "workplace_address_street_label": null,
    "workplace_address_zipcode": "08000",
    "workplace_brand": null,
    "workplace_description": null,
    "workplace_geopoint": null,
    "workplace_idcc": null,
    "workplace_legal_name": null,
    "workplace_naf_code": null,
    "workplace_naf_label": null,
    "workplace_name": "GROUPE GCA",
    "workplace_opco": null,
    "workplace_siret": null,
    "workplace_size": null,
    "workplace_website": null,
  },
]
`;
